typed-racket-lib/typed-racket/rep/base-type-rep.rkt

```diff
@@ -0,0 +1,109 @@
+#lang racket/base
+
+(require "rep-utils.rkt"
+         "core-rep.rkt"
+         "type-mask.rkt"
+         racket/match
+         (for-syntax racket/base
+                     racket/syntax
+                     syntax/parse))
+
+(provide define-base-types
+         Base?
+         Base-name
+         Base-predicate
+         Base-bits:
+         Base-predicate:
+         Base-name/contract:)
+
+;;-----------------
+;; Base Type
+;;-----------------
+
+;; name - a printable type name (unique for each Base)
+;; numeric? - is this Base type for values where 'number?' is #t
+;; bits - the binary representation of this Base type, used in BaseUnions
+;; to efficiently represent unions of Bases. NOTE: this should include only
+;; one bit set to 1.
+;; contract - used when generating contracts from types
+;; predicate - used to check (at compile-time) whether a value belongs
+;; to that base type. This is used to check for subtyping between value
+;; types and base types.
+(def-type Base ([name symbol?]
+                [numeric? boolean?]
+                [bits exact-nonnegative-integer?]
+                [contract syntax?]
+                [predicate procedure?])
+  #:base
+  #:no-provide
+  [#:mask (λ (t) (if (Base-numeric? t)
+                     mask:number
```

Comment:
if we expanded the range to go beyond 30 bits we could go nuts here =) (and I bet the diff between `bitwise-...` and `unsafe-fx...` is not noticable big picture wise